doc/source/workflows/checkpointing.rst

```diff
@@ -0,0 +1,257 @@
+
+Besides skipping checkpoints, an alternative is to replace workflow steps with Ray remote functions. But there are 3 reasons against it:
+
+1. You cannot compose a workflow DAG by mixing workflow steps and Ray remote functions.
+2. Unlike workflow steps, Ray workflow would not display progress and status of Ray remote functions.
+3. The remote functions won’t be considered as part of the workflow, so they won’t be recovered automatically when they fail. For example, instead of returning a workflow (step) B in workflow step A, you return a ObjectRef of a remote function. With workflow step B, workflow enables recovery from step B; but with the remote function, the workflow can only recover from step A.
+
+
+Semantics of Checkpointing
+--------------------------
+
+Understand Checkpoints
+~~~~~~~~~~~~~~~~~~~~~~
+
+By default, a workflow step checkpoints its output when finished execution. The typical output is some values returned by a workflow step, for example:
+
+
+.. code-block:: python
+
+    @workflow.step
+    def read_data(num: int):
+        return [i for i in range(num)]
+        assert handler.double.run(True) == 4
+
+
+But the output could be more general than a value, this is because a parent workflow may return a child workflow. In the example below, ``book_trip`` returns ``finalize_or_cancel``, which implicitly contains a workflow DAG including all steps inside, their inputs and DAG structure etc. So the output of ``book_trip`` is the workflow DAG inside it, and the output is checkpointed when ``book_trip`` is returned.
+
+.. code-block:: python
+
+    @workflow.step
+    def book_trip(origin: str, dest: str, dates) -> "Workflow[Receipt]":
+        # Note that the workflow engine will not begin executing
+        # child workflows until the parent step returns.
+        # This avoids step overlap and ensures recoverability.
+        f1: Workflow = book_flight.step(origin, dest, dates[0])
+        f2: Workflow = book_flight.step(dest, origin, dates[1])
+        hotel: Workflow = book_hotel.step(dest, dates)
+        return finalize_or_cancel.step([f1, f2], [hotel])
+
+Skipping checkpointing the DAG inside ``book_trip`` does not mean we cannot recover from a checkpoint produced “inside” ``book_trip``. In this case, ``book_trip`` returns ``finalize_or_cancel``, so if ``finalize_or_cancel`` returns and checkpointed its output (a python value of another DAG), when resuming ``book_trip``, we can resume from this output checkpoint.
+
+For a top-level workflow (e.g., a workflow in a driver script), it's DAG will be ``checkpointed`` before execution. We can imagine there is an invisible “root” workflow step, where the top-level workflow is its output. This also implies we cannot skip checkpointing the DAG of the top-level workflow, which makes sense (otherwise we cannot recover from the failure of the workflow).
+
```

Comment:
I think this description may actually make it more confusing. I would think about just introducing the API right away and then showing the difference between checkpointing values and sub-workflows in the examples.